File: drake_ros_bazel_installed/tools/ros2/test/isolated_listener.cc
#include <chrono>
#include <memory>

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include "tools/ros2/rmw_isolation.h"

using std::placeholders::_1;

class IsolatedListener : public rclcpp::Node
{
  public:
    IsolatedListener()
    : rclcpp::Node("isolated_listener")
    {
      double timeout = this->declare_parameter("timeout", 2.0);
      subscription_ = this->create_subscription<std_msgs::msg::String>(
          "topic", 10, std::bind(&IsolatedListener::topic_callback, this, _1));
      timer_ = this->create_wall_timer(
          std::chrono::duration<double>(timeout),
          std::bind(&IsolatedListener::timer_callback, this));
    }

  private:
    void topic_callback(const std_msgs::msg::String::SharedPtr msg) const
    {
      throw std::runtime_error("I heard '" + msg->data + "'");
    }

    void timer_callback() const
    {
      rclcpp::shutdown();
    }

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  const char * TEST_TMPDIR = std::getenv("TEST_TMPDIR");
  if (TEST_TMPDIR != nullptr) {
    ros2::isolate_rmw_by_path(argv[0], TEST_TMPDIR);
  }
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<IsolatedListener>());
  rclcpp::shutdown();
  return 0;
}

File: drake_ros_bazel_installed/tools/ros2/test/isolated_listener.py
import os

import rclpy
import rclpy.node
import rclpy.executors
import std_msgs.msg


class IsolatedListener(rclpy.node.Node):

    def __init__(self):
        super().__init__('isolated_listener')
        self._subscription = self.create_subscription(
            std_msgs.msg.String,
            'topic',
            self._topic_callback,
            10)
        timeout = self.declare_parameter('timeout', 2.0)
        self._timer = self.create_timer(
            timeout.value, self._timer_callback)

    def _topic_callback(self, msg):
        assert False, f"I heard '{msg.data}'!"

    def _timer_callback(self):
        rclpy.shutdown()

if __name__ == '__main__':
    if 'TEST_TMPDIR' in os.environ:
        from drake_ros.tools.ros2.rmw_isolation import isolate_rmw_by_path
        isolate_rmw_by_path(os.environ['TEST_TMPDIR'])

    rclpy.init()

    try:
        executor = rclpy.executors.SingleThreadedExecutor()
        rclpy.spin(IsolatedListener(), executor)
    finally:
        # NOTE(hidmic): try_shutdown raises AttributeError
        # Need https://github.com/ros2/rclpy/pull/812
        pass  # rclpy.try_shutdown()
